core/constraints/validation.go

```diff
@@ -215,20 +219,79 @@ func (v *validator) checkValidValues(cons Value) error {
 	return nil
 }
 
+// InvalidVocabValueErr represents an error that occurs
+// when a validation constraint is violated.
+// It provides details for the
+// invalid inputs, closest valid values and all possible valid values.
+type InvalidVocabValueErr struct {
+	closestValidValues  []string
+	allValidValues      []string
+	inputAttributeName  string
+	inputAttributeValue any
+}
+
+func (ve *InvalidVocabValueErr) Error() string {
+	closestValuesStr := strings.Join(ve.closestValidValues, " ")
+	errStr := fmt.Sprintf("invalid constraint value: %v=%v\nvalid values are: %v",
+		ve.inputAttributeName, ve.inputAttributeValue, closestValuesStr)
+
+	// Add additional items if there are more than the validValuesCountLimit
+	additionalItemsCount := len(ve.allValidValues) - len(ve.closestValidValues)
+	if additionalItemsCount > 0 {
+		return fmt.Sprintf("%s ...(plus %d more)", errStr, additionalItemsCount)
+	}
+
+	return errStr
+}
+
 // checkInVocab returns an error if the attribute value is not allowed by the
 // vocab which may have been registered for it.
-func (v *validator) checkInVocab(attributeName string, attributeValue interface{}) error {
+func (v *validator) checkInVocab(attributeName string, attributeValue interface{}) *InvalidVocabValueErr {
```

Comment:
Do you need the return type to be `*InvalidVocabValueErr`? This should probably just be `error`.

Reply:
Yup not a requirement, sure updated with thanks Harry :)